test/framework/ownerreference_helpers.go

```diff
@@ -0,0 +1,308 @@
+	setClusterPause(ctx, proxy.GetClient(), clusterKey, true)
+
+	removeOwnerReferences(ctx, proxy, namespace)
+
+	// Unpause the cluster by setting .spec.paused: "false"
+	setClusterPause(ctx, proxy.GetClient(), clusterKey, false)
+
+	// Annotate the clusterClass, if one is in use, to speed up reconciliation.
+	annotateClusterClass(ctx, proxy.GetClient(), clusterKey)
+
+	// Check that the ownerReferences are as expected after additional reconciliations.
+	AssertOwnerReferences(namespace, proxy.GetKubeconfigPath(), assertFuncs...)
+}
+
+func AssertOwnerReferences(namespace, kubeconfigPath string, assertFuncs ...map[string]func(reference []metav1.OwnerReference) error) {
+	allAssertFuncs := map[string]func(reference []metav1.OwnerReference) error{}
+	for _, m := range assertFuncs {
+		for k, v := range m {
+			allAssertFuncs[k] = v
+		}
+	}
+	Eventually(func() error {
+		graph, err := clusterctlcluster.GetOwnerGraph(namespace, kubeconfigPath)
+		Expect(err).To(BeNil())
+		allErrs := []error{}
+		for _, v := range graph {
+			if _, ok := allAssertFuncs[v.Object.Kind]; !ok {
+				allErrs = append(allErrs, fmt.Errorf("kind %s does not have an associated ownerRef assertion function", v.Object.Kind))
+				continue
+			}
+			if err := allAssertFuncs[v.Object.Kind](v.Owners); err != nil {
+				allErrs = append(allErrs, errors.Wrapf(err, "Unexpected ownerReferences for %s/%s", v.Object.Kind, v.Object.Name))
+			}
+		}
+		var errStrings string
+		for _, err := range allErrs {
+			errStrings = fmt.Sprintf("%s\n%s", errStrings, err.Error())
+		}
+		return kerrors.NewAggregate(allErrs)
+	}).Should(Succeed())
+}
+
+// Kind and GVK for types in the core API package.
+var (
+	clusterClassKind       = "ClusterClass"
+	clusterKind            = "Cluster"
+	machineKind            = "Machine"
+	machineSetKind         = "MachineSet"
+	machineDeploymentKind  = "MachineDeployment"
+	machineHealthCheckKind = "MachineHealthCheck"
+
+	clusterClassGVK      = clusterv1.GroupVersion.WithKind(clusterClassKind)
+	clusterGVK           = clusterv1.GroupVersion.WithKind(clusterKind)
+	machineDeploymentGVK = clusterv1.GroupVersion.WithKind(machineDeploymentKind)
+	machineSetGVK        = clusterv1.GroupVersion.WithKind(machineSetKind)
+	machineGVK           = clusterv1.GroupVersion.WithKind(machineKind)
+)
+
+// CoreTypeOwnerReferenceAssertion maps Cluster API core types to functions which return an error if the passed
+// OwnerReferences aren't as expected.
+var CoreTypeOwnerReferenceAssertion = map[string]func([]metav1.OwnerReference) error{
+	clusterClassKind: func(owners []metav1.OwnerReference) error {
+		return hasExactOwnersByGVK(owners, []schema.GroupVersionKind{})
+	},
+	clusterKind: func(owners []metav1.OwnerReference) error {
+		return hasExactOwnersByGVK(owners, []schema.GroupVersionKind{})
+	},
+	machineDeploymentKind: func(owners []metav1.OwnerReference) error {
+		return hasExactOwnersByGVK(owners, []schema.GroupVersionKind{clusterGVK})
```

Comment:
```suggestion
                 // Machine deployments must be owned by the Cluster.
		return hasExactOwnersByGVK(owners, []schema.GroupVersionKind{clusterGVK})
```

and so on for the other kinds.